### PR TITLE
Add in-memory functional test project

### PR DESCRIPTION
**Pull request by halter73**
- Run AddressRegistrationTests independently to avoid port conflicts
- Only run tests that verify transport behavior for each transport
- Improve file logging for some flaky tests

#2635

### CODE REVIEW

**Comment by Tratcher** [comment]
Max Path issues on OSX?

File: test/Kestrel.InMemory.FunctionalTests/TestTransport/InMemoryTransportConnection.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        private readonly CancellationTokenSource _connectionClosedTokenSource = new CancellationTokenSource();
+
+        private bool _isClosed;
+
+        public InMemoryTransportConnection(MemoryPool<byte> memoryPool)
+        {
+            MemoryPool = memoryPool;
+
+            LocalAddress = IPAddress.Loopback;
+            RemoteAddress = IPAddress.Loopback;
```
[the remark this answers is not in the document]
Comment (halter73) [reply test/Kestrel.InMemory.FunctionalTests/TestTransport/InMemoryTransportConnection.cs]: The ports default to "0".

**Comment by JunTaoLuo** [comment]
Can we backport this to 2.1 as well? Some of the tests are still flaky there.

**Comment by halter73** [comment]
Sounds fine to me.

**Pull request merged**